Fix deferred object type mismatch

falsandtru opened the pull request:
# ~~Generics~~

## ~~T => any~~
Current definition is `T => any`.
Fixed definition is `T => T`.

## ~~T[] => any[]~~
Current definition is `T, ...T[] => T, ...any[]` and `...T[] => ...any[]`.
Fixed definition is `...T[] => ...T[]`.

# Parameter
Current definition is `(T[], ...T[])`.
Fixed definition is `(T[], ...T[][])`.

# Massaging callback type
Current definition operatable messaging value in `done/fail/always` methods.
Fixed definition unoperatable messaging value.

# Messaging return type
Current `progress/done/fail/always` methods definition return promise type.
Fixed to return deferred type.

# Source
http://api.jquery.com/category/deferred-object/

falsandtru commented:
now fixing...

falsandtru commented:
Adjusted in current `done` method definition

```ts
done(doneCallbacks1?: JQueryPromiseCallback<T>, ...doneCallbacks2: any[]): JQueryDeferred<T>;
```

and `resolve` method definition

```ts
resolve(value?: T, ...args: any[]): JQueryDeferred<T>;`
```

falsandtru commented:
# Method

Current progress/done/fail/always method definition return promise type.
Fixed to return deferred type.

falsandtru commented:
# Source
http://api.jquery.com/category/deferred-object/

vvakame commented:
@johnnyreilly

johnnyreilly commented:
Thanks for this @falsandtru.  Pretty much looks good.  Could you explain the removal of `done`, `always`, `fail` and `progress` from `JQueryDeferred` please?

I didn't quite follow the explanation you gave...  SInce the tests pass I'm inclined to merge but I'd like to understand this a little better if possible first.  Thanks.

falsandtru commented:
Hi @johnnyreilly.
Sorry, my english power is poor. Are you sure you want if explain why I remove these methods?

Removed methods `progress/done/fail/always` is a duplicate definition. These methods type is the same for all locations. And these Promise methods is subset of Deferred methods. Promise type definition will be merged into the Deferred by inheritance.

> http://api.jquery.com/Types/#Promise
Promise Object
This object provides a **subset of the methods of the Deferred object (then, done, fail, always, pipe, progress, and state)** to prevent users from changing the state of the Deferred.

And, callbacks and return values type is different.

> http://api.jquery.com/deferred.done/
deferred.done( doneCallbacks [, doneCallbacks ] )Returns: **Deferred**
...
doneCallbacks
Type: Function()
A **function, or array of functions**, that are called when the Deferred is resolved.

It should be noted, I reverted the all causes of first errors. Thanks.

johnnyreilly commented:
That's great - just wanted to understand.  Thanks for your work :+1:
